kernel/src/mmio.rs

```diff
@@ -0,0 +1,154 @@
+use core::{
+    mem::size_of,
+    ops::{Deref, DerefMut},
+};
+
+use crate::{
+    arch_api::{asm::memory_barrier, paging::MemoryType},
+    heap::{map_physical_memory, PhysicalAddressHandle},
+};
+
+#[derive(Copy, Clone)]
+pub struct MmioPointer<T> {
+    ptr: *mut T,
+}
+
+impl<T> MmioPointer<T> {
+    pub fn new(ptr: *mut T) -> Self {
+        Self { ptr }
+    }
+
+    /// # Safety
+    /// Since this function is essentially a wrapper around a raw pointer, it is inherently unsafe.
+    /// Additionally, MMIO accesses may have side effects, so this function is unsafe for that reason as well.
+    pub unsafe fn read(&self) -> T {
+        let result = unsafe { self.ptr.read_volatile() };
+        memory_barrier();
+        result
+    }
+
+    /// # Safety
+    /// Since this function is essentially a wrapper around a raw pointer, it is inherently unsafe.
+    /// Additionally, MMIO accesses may have side effects, so this function is unsafe for that reason as well.
+    pub unsafe fn write(&mut self, value: T) {
+        unsafe { self.ptr.write_volatile(value) };
+        memory_barrier();
+    }
+}
+
+pub struct MmioRange {
+    start: *mut u8,
+    size: usize,
+}
+
+impl MmioRange {
+    pub fn new(start: *mut u8, size: usize) -> Self {
+        Self { start, size }
+    }
+
+    /// # Safety
+    /// The same rules apply as for performing arithmetic on a raw pointer.
+    pub unsafe fn at_offset<T>(&self, offset: usize) -> MmioPointer<T> {
+        assert!(offset + size_of::<T>() <= self.size);
+        MmioPointer::new(unsafe { self.start.add(offset) as *mut T })
+    }
+
+    pub fn resize(&mut self, new_size: usize) {
+        self.size = new_size;
+    }
+
+    pub fn size(&self) -> usize {
+        self.size
+    }
+}
+
+pub struct MmioMemoryHandle {
+    physical_memory_handle: PhysicalAddressHandle,
+    mmio_range: MmioRange,
+}
+
+impl MmioMemoryHandle {
+    /// # Safety
+    /// see `heap::map_physical_memory`.
+    pub unsafe fn new(physical_address: usize, size: usize) -> Self {
+        let mut handle = map_physical_memory(physical_address, size, MemoryType::Device);
+        let mmio_range = MmioRange::new(
+            PhysicalAddressHandle::as_mut_ptr(&mut handle),
+            PhysicalAddressHandle::size(&handle),
+        );
+        Self {
+            physical_memory_handle: handle,
+            mmio_range,
+        }
+    }
+}
+
+impl Deref for MmioMemoryHandle {
+    type Target = MmioRange;
+
+    fn deref(&self) -> &Self::Target {
+        &self.mmio_range
+    }
+}
+
+impl DerefMut for MmioMemoryHandle {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.mmio_range
+    }
+}
+
+#[cfg(test)]
```

Comment:
So awesome to see these tests. Not only does it prevent regression but it also documents how the code is expected to be used. Very nice.